core/primitives/src/runtime/config_store.rs

```diff
@@ -0,0 +1,153 @@
+use crate::runtime::config::RuntimeConfig;
+use crate::types::ProtocolVersion;
+use std::collections::BTreeMap;
+use std::iter::FromIterator;
+use std::ops::Bound;
+use std::sync::Arc;
+
+macro_rules! include_config {
+    ($file:expr) => {
+        include_bytes!(concat!("../../../../nearcore/res/runtime_configs/", $file))
+    };
+}
+
+/// Stores pairs of protocol versions for which runtime config was updated and
+/// the new runtime config in bytes.
+/// Protocol versions are given in increasing order. First one is always 0, so that each version is
+/// mapped to some config.
+static CONFIGS: [(ProtocolVersion, &[u8]); 2] =
+    [(0, include_config!("29.json")), (42, include_config!("42.json"))];
+
+/// Stores runtime config for each protocol version where it was updated.
+#[derive(Debug)]
+pub struct RuntimeConfigStore {
+    /// Maps protocol version to the config.
+    store: BTreeMap<ProtocolVersion, Arc<RuntimeConfig>>,
+}
+
+impl RuntimeConfigStore {
+    /// Constructs a new store.
+    ///
+    /// If genesis_runtime_config is Some, configs for protocol versions 0 and 42 are overridden by
+    /// this config and config with lowered storage cost, respectively.
+    /// This is done to preserve compatibility with previous implementation, where we updated
+    /// runtime config by sequential modifications to the genesis runtime config.
+    /// TODO #4775: introduce new protocol version to have the same runtime config for all chains
+    pub fn new(genesis_runtime_config: Option<&RuntimeConfig>) -> Self {
+        let mut store =
+            BTreeMap::from_iter(CONFIGS.iter().cloned().map(|(protocol_version, config_bytes)| {
+                (protocol_version, Arc::new(serde_json::from_slice(config_bytes).unwrap()))
+            }));
+
+        if let Some(runtime_config) = genesis_runtime_config {
+            let mut config = runtime_config.clone();
+            store.insert(0, Arc::new(config.clone()));
+
+            config.storage_amount_per_byte = 10u128.pow(19);
+            store.insert(42, Arc::new(config.clone()));
+        }
```

Comment:
This piece of code is a bit confusing and hard to read, but I also don't see a much better way to do it. I guess we dug a hole for ourselves :(